C# 7.x: Default literals (#236)
dotnet/csharpstandard: standard/expressions.md

### 11.7.19 Default value expressions

> A default value expression is used to obtain the default value ([§9.3](variables.md#93-default-values)) of an explicit or inferred type. Typically, a default value expression is used for type parameters, since it might not be known if the type parameter is a value type or a reference type. (No conversion exists from the `null` literal ([§6.4.5.7](lexical-structure.md#6457-the-null-literal)) to a type parameter unless the type parameter is known to be a reference type ([§8.2](types.md#82-reference-types)).)

Reviewer: Since the introduction of the default literal, I'm not sure that this "typically" is correct. For example, I often see it used for default parameter values, especially for CancellationToken:

```csharp
public void DoSomethingAsync(CancellationToken cancellationToken = default)
```

No type parameters involved. Would "One notable use" be a suitable substitution here? And should everything after the first sentence be in a note?

Reviewer 2: I agree. It's not normative. §10.2.12 and §10.3.8 have normative text that there isn't a conversion from `null` to a type parameter that isn't known to be a reference type.

Reviewer 3: I don't think the phrasing "an explicit or inferred type" is correct here (and elsewhere), we should follow the phraseology used for `null`. So something along the lines of "obtain the default value of a [target] type" depending on the context.

Reviewer 2: I agree with Nigel's point on "inferred type". When I read that, I think it would allow this:

```csharp
var i = default;
```

That's not allowed. I think the following suggestion on "conversion to any type" would be better: it would work for both known types and type parameters.

```ANTLR
default_value_expression
    : explictly_typed_default
    | default_literal
    ;

explictly_typed_default
    : 'default' '(' type ')'
    ;

default_literal
    : 'default'
    ;
```

> *default_literal* allows the type to be inferred by target-typing.

Reviewer: I don't think "target-typing" is a phrase we use anywhere else. Should we drop this sentence entirely? I don't see anything about *how* the type of a default literal is inferred, but maybe that's already covered in general? (I'm genuinely unsure about that.)

> If the explicit or inferred *type* in a *default_value_expression* evaluates at run-time to a reference type, or in an unsafe context, to a pointer type ([§22.3](unsafe-code.md#223-pointer-types)), the result is `null` converted to that type. If the explicit or inferred *type* in a *default_value_expression* evaluates at run-time to a value type, the result is the value type's default value ([§8.3.3](types.md#833-default-constructors)).

Reviewer: I don't think `*type*` should be in italics here (either time), given that it's only applicable in one branch of *default_value_expression*.

> If the explicit or inferred *type* in a *default_value_expression* evaluates at run-time to a reference type, or in an unsafe context, to a pointer type ([§22.3](unsafe-code.md#223-pointer-types)), the result is `null` converted to that type. If the explicit or inferred *type* in a *default_value_expression* evaluates at run-time to a value type, the result is the value type's default value ([§8.3.3](types.md#833-default-constructors)).

Reviewer: ```suggestion
If the explicit or inferred type in a *default_value_expression* evaluates at run-time to a reference type, or in an unsafe context, to a pointer type ([§22.3](unsafe-code.md#223-pointer-types)), the result is `null` converted to that type. If the explicit or inferred type in a *default_value_expression* evaluates at run-time to a value type, the result is the value type's default value ([§8.3.3](types.md#833-default-constructors)).
```

> If the explicit or inferred *type* in a *default_value_expression* evaluates at run-time to a reference type, or in an unsafe context, to a pointer type ([§22.3](unsafe-code.md#223-pointer-types)), the result is `null` converted to that type. If the explicit or inferred *type* in a *default_value_expression* evaluates at run-time to a value type, the result is the value type's default value ([§8.3.3](types.md#833-default-constructors)).

Reviewer: This duplicates information in 9.3, which is entitled "default values" but is about default values of variables.

I wonder whether it might be worth refactoring to have "default value of a type" somewhere in clause 8 (types) and then refer to that from both here and 9.3. That can come later though.

Reviewer 3: Agree with @jskeet. This duplicates information in §9.3 and in the first line of this clause. Just remove the whole paragraph. If we need to talk about *runtime type* either incorporate that into the first line of the clause or replace this paragraph with one saying the value produced depends on the target type at runtime – but that might be implicit elsewhere?

Reviewer: Doing this in a suggestion. (Runtime type is basically about type parameters, but doesn't need to be called out if we're referring to 9.3.)

> A *default_value_expression* is a constant expression ([§11.20](expressions.md#1120-constant-expressions)) if explicit or inferred *type* is a reference type or a type parameter that is known to be a reference type ([§8.2](types.md#82-reference-types)). In addition, a *default_value_expression* is a constant expression if the type is one of the following value types: `sbyte`, `byte`, `short`, `ushort`, `int`, `uint`, `long`, `ulong`, `char`, `float`, `double`, `decimal`, `bool,` or any enumeration type.

Reviewer: ```suggestion
A *default_value_expression* is a constant expression ([§11.20](expressions.md#1120-constant-expressions)) if the explicit or inferred type is a reference type or a type parameter that is known to be a reference type ([§8.2](types.md#82-reference-types)). In addition, a *default_value_expression* is a constant expression if the type is one of the following value types: `sbyte`, `byte`, `short`, `ushort`, `int`, `uint`, `long`, `ulong`, `char`, `float`, `double`, `decimal`, `bool,` or any enumeration type.
```

### 11.7.19 Default value expressions

> A default value expression is used to obtain the default value ([§9.3](variables.md#93-default-values)) of an explicit or inferred type. Typically, a default value expression is used for type parameters, since it might not be known if the type parameter is a value type or a reference type. (No conversion exists from the `null` literal ([§6.4.5.7](lexical-structure.md#6457-the-null-literal)) to a type parameter unless the type parameter is known to be a reference type ([§8.2](types.md#82-reference-types)).)

Reviewer: ```suggestion
A default value expression is used to obtain the default value ([§9.3](variables.md#93-default-values)) of a type.
```

```ANTLR
default_value_expression
    : explictly_typed_default
    | default_literal
    ;

explictly_typed_default
    : 'default' '(' type ')'
    ;

default_literal
    : 'default'
    ;
```

> A *default_literal* represents a default value (§9.3). It does not have a type, but can be converted to any type through a default literal conversion (§default-literal-new-clause).

Reviewer: ```suggestion
A *default_literal* represents a default value ([§9.3](variables.md#93-default-values)). It does not have a type, but can be converted to any type through a default literal conversion (§default-literal-new-clause).
```

> *default_literal* allows the type to be inferred by target-typing.

Reviewer 3: Agree with @jskeet, "target-typing" is not used in the Standard. I think we should describe *default_literal* in the same way as a *null_literal* is described, i.e. replace this line with:

> A `default_literal` represents a default value (§9.3). It does not have a type, but can be converted to any type
or nullable through a default literal conversion (§default-literal-new-clause)

Reviewer: That looks good to me. Will discuss at the meeting later "today" (UTC). @Nigel-Ecma: I know you can't make the meeting, but I'm going to assume you're happy with your own suggestion.

Reviewer 3: @jskeet – in this rare case I do agree with myself ;-)

> *default_literal* allows the type to be inferred by target-typing.

Reviewer: ```suggestion
A *default_literal* represents a default value (§9.3). It does not have a type, but can be converted to any type through a default literal conversion (§default-literal-new-clause).
```

Reviewer: @Nigel-Ecma: Removed the nullable part as "a type" is fine. (In the null literal it's to a reference type or nullable value type; we don't need the restriction here.)

> If the explicit or inferred *type* in a *default_value_expression* evaluates at run-time to a reference type, or in an unsafe context, to a pointer type ([§22.3](unsafe-code.md#223-pointer-types)), the result is `null` converted to that type. If the explicit or inferred *type* in a *default_value_expression* evaluates at run-time to a value type, the result is the value type's default value ([§8.3.3](types.md#833-default-constructors)).

Reviewer: ```suggestion
The result of a *default_value_expression* is the default ([§9.3](variables.md#93-default-values)) of the explicit type in an *explictly_typed_default*, or the target type of the conversion for a *default_value_expression*.
```

> If the explicit or inferred *type* in a *default_value_expression* evaluates at run-time to a reference type, or in an unsafe context, to a pointer type ([§22.3](unsafe-code.md#223-pointer-types)), the result is `null` converted to that type. If the explicit or inferred *type* in a *default_value_expression* evaluates at run-time to a value type, the result is the value type's default value ([§8.3.3](types.md#833-default-constructors)).

Reviewer 2: Instead of "inferred type", I would use the conversion language proposed earlier. Also, I don't think this happens *at run time* so I removed that phrase.

```suggestion
If the *default_value_expression* is converted to a reference type, or in an unsafe context, to a pointer type ([§22.3](unsafe-code.md#223-pointer-types)), the result is `null` converted to that type. If the *default_value_expression* is converted to a value type, the result is the value type's default value ([§8.3.3](types.md#833-default-constructors)).
```

> A *default_value_expression* is a constant expression ([§11.21](expressions.md#1121-constant-expressions)) if explicit or inferred *type* is a reference type or a type parameter that is known to be a reference type ([§8.2](types.md#82-reference-types)). In addition, a *default_value_expression* is a constant expression if the type is one of the following value types: `sbyte`, `byte`, `short`, `ushort`, `int`, `uint`, `long`, `ulong`, `char`, `float`, `double`, `decimal`, `bool,` or any enumeration type.

Reviewer: ```suggestion
A *default_value_expression* is a constant expression ([§11.21](expressions.md#1121-constant-expressions)) if the type is one of:

- a reference type
- a type parameter that is known to be a reference type ([§8.2](types.md#82-reference-types));
- one of the following value types: `sbyte`, `byte`, `short`, `ushort`, `int`, `uint`, `long`, `ulong`, `char`, `float`, `double`, `decimal`, `bool,`; or
- any enumeration type.
```

> A *default_value_expression* is a constant expression ([§11.21](expressions.md#1121-constant-expressions)) if explicit or inferred *type* is a reference type or a type parameter that is known to be a reference type ([§8.2](types.md#82-reference-types)). In addition, a *default_value_expression* is a constant expression if the type is one of the following value types: `sbyte`, `byte`, `short`, `ushort`, `int`, `uint`, `long`, `ulong`, `char`, `float`, `double`, `decimal`, `bool,` or any enumeration type.

Reviewer 3: Use of "explicit or inferred" to remove. The "In addition" phrasing seems wrong, I'd be tempted to just use a single list:

> A *default_value_expression* is a constant expression ([§11.21](expressions.md#1121-constant-expressions)) if the type is:
> - a reference type or a type parameter that is known to be a reference type ([§8.2](types.md#82-reference-types));
> - one of the following value types: `sbyte`, `byte`, `short`, `ushort`, `int`, `uint`, `long`, `ulong`, `char`, `float`, `double`, `decimal`, `bool,`; or
> - any enumeration type.

Reviewer: Resolving this comment having added it as a suggestion.
